File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

def main():
    n = int(input())
    stack = list()
    for i in range(n):
        result = True
        s = input().strip()
        for j in range(len(s)):
            stack.append( s[ j ] )
            
            if stack[-1] == ")":
                stack.pop(-1)
                if not stack:
                    result = False
                    break
                stack.pop(-1)
        if stack or not result:
            print("NO")
            stack.clear()
        elif not stack:
            print("YES")             
main()

File: 백준/Silver/9012. 괄호/괄호.cc
// Authored by : chjh2129
#include <bits/stdc++.h>
using namespace std;

int t;
string s;

// 짝이 맞으면 true, 틀리면 false를 반환
bool solve(){
  stack<char> stk;
  for(auto c : s){
    if(c == '(') stk.push(c);
    else if(stk.empty() || stk.top() != '(') return 0;
    else stk.pop();
  }
  return stk.empty();
} 

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> t;
  while(t--){
    cin >> s;
    // solve & output
    cout << (solve() ? "YES\n" : "NO\n");
  }
}